targets/TARGET_Cypress/TARGET_PSOC6/i2c_api.c

```diff
@@ -0,0 +1,692 @@
+        case CY_SCB_I2C_MASTER_MANUAL_ADDR_NAK:
+        case CY_SCB_I2C_MASTER_MANUAL_NAK:
+            return I2C_ERROR_NO_SLAVE;
+
+        case CY_SCB_I2C_SUCCESS:
+        case CY_SCB_I2C_BAD_PARAM:
+        default:
+            return 0;
+    }
+ }
+
+/**
+ * Callback function to handle into and out of deep sleep state transitions.
+ */
+#if DEVICE_SLEEP && DEVICE_LOWPOWERTIMER
+static cy_en_syspm_status_t i2c_pm_callback(cy_stc_syspm_callback_params_t *callback_params)
+{
+    cy_stc_syspm_callback_params_t params = *callback_params;
+    i2c_obj_t *obj = (i2c_obj_t *)params.context;
+    params.context = &obj->context;
+
+    return Cy_SCB_I2C_DeepSleepCallback(&params);
+}
+#endif // DEVICE_SLEEP && DEVICE_LOWPOWERTIMER
+
+
+/** Initialize the I2C peripheral. It sets the default parameters for I2C
+ *  peripheral, and configures its specified pins.
+ *
+ *  @param obj  The I2C object
+ *  @param sda  The sda pin
+ *  @param scl  The scl pin
+ */
+void i2c_init(i2c_t *obj_in, PinName sda, PinName scl)
+{
+    i2c_obj_t *obj = OBJ_P(obj_in);
+    uint32_t i2c = pinmap_peripheral(sda, PinMap_I2C_SDA);
+    i2c = pinmap_merge(i2c, pinmap_peripheral(scl, PinMap_I2C_SCL));
+    if (i2c != (uint32_t)NC) {
+        if (cy_reserve_io_pin(sda) || cy_reserve_io_pin(scl)) {
+            error("I2C pin reservation conflict.");
+        }
+        obj->base = (CySCB_Type*)i2c;
+        obj->i2c_id = ((I2CName)i2c - I2C_0) / (I2C_1 - I2C_0);
+        obj->pin_sda = sda;
+        obj->pin_scl = scl;
+        obj->clock = CY_PIN_CLOCK(pinmap_function(scl, PinMap_I2C_SCL));
+        obj->divider = I2C_INVALID_DIVIDER;
+        obj->mode = CY_SCB_I2C_MASTER;
+        obj->timeout = I2C_DEFAULT_TIMEOUT;
+#if DEVICE_I2C_ASYNCH
+        obj->pending = PENDING_NONE;
+        obj->events = 0;
+#endif // DEVICE_I2C_ASYNCH
+        i2c_init_clock(obj, I2C_DEFAULT_SPEED);
+        i2c_init_pins(obj);
+        i2c_init_peripheral(obj);
+#if DEVICE_SLEEP && DEVICE_LOWPOWERTIMER
+		obj->pm_callback_handler.callback = i2c_pm_callback;
+		obj->pm_callback_handler.type = CY_SYSPM_DEEPSLEEP;
+		obj->pm_callback_handler.skipMode = 0;
+		obj->pm_callback_handler.callbackParams = &obj->pm_callback_params;
+		obj->pm_callback_params.base = obj->base;
+		obj->pm_callback_params.context = obj;
+		if (!Cy_SysPm_RegisterCallback(&obj->pm_callback_handler)) {
+			error("PM callback registration failed!");
+		}
+#endif // DEVICE_SLEEP && DEVICE_LOWPOWERTIMER
+    } else {
+        error("I2C pinout mismatch. Requested pins Rx and Tx can't be used for the same I2C communication.");
+    }
+}
+
+/** Configure the I2C frequency
+ *
+ *  @param obj The I2C object
+ *  @param hz  Frequency in Hz
+ */
+void i2c_frequency(i2c_t *obj_in, int hz)
+{
+    i2c_obj_t *obj = OBJ_P(obj_in);
+    Cy_SCB_I2C_Disable(obj->base, &obj->context);
+    i2c_init_clock(obj, hz);
+    Cy_SCB_I2C_Enable(obj->base);
+}
+
+/** Send START command
+ *
+ *  This function is a dummy operation on PSoC
+ */
```

Comment:
In this case, having this is OK as it is different from i2c_start documentation and explains why this is empty (regarding my earlier comment about having copy-paste API docs in the implementation)